### PR TITLE
JWT Authentication for API Gateway

### PR DESCRIPTION
**Pull request by ChamodDamitha**
Enable API authentication support using JWT tokens.

Issue - https://github.com/wso2/product-apim/issues/5115

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+    @MethodStats
+    public AuthenticationContext authenticate(String jwtToken, MessageContext synCtx, OpenAPI openAPI)
+            throws APISecurityException {
+
+        String[] splitToken = jwtToken.split("\\.");
+
+        JSONObject payload = null;
+        boolean isVerified = false;
+
+        String tokenSignature = splitToken[2];
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java]: possible  array index out of bounds if someone sends a invalid jwt?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+     *
+     * @param splitToken The JWT token which is split into [header, payload, signature]
+     * @return whether the signature algorithm
+     * @throws APISecurityException in case of signature algorithm extraction failure
+     */
+    private String getSignatureAlgorithm(String[] splitToken) throws APISecurityException {
+        String signatureAlgorithm;
+        JSONObject header;
+        try {
+            header = new JSONObject(new String(Base64.getUrlDecoder().decode(splitToken[0])));
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java]: seems we do url decode? Can there be any other decoding methods?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+        } catch (JSONException | IllegalArgumentException e) {
+            log.debug("Token decryption failure when retrieving header.", e);
+            throw new APISecurityException(APISecurityConstants.API_AUTH_INVALID_CREDENTIALS,
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java]: saying token invalid. Can't we trim and log atkeast last few letters

File: components/apimgt/org.wso2.carbon.apimgt.gateway/pom.xml [hunk 4]
```diff
@@ -231,11 +231,15 @@
             <groupId>org.wso2.carbon.mediation</groupId>
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/pom.xml]: Should we validate th signature of the token via JWKS endpoint?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/pom.xml [hunk 5]
```diff
@@ -345,6 +349,7 @@
                             org.apache.synapse.endpoints.*,
                             org.apache.synapse.mediators.base,
                             org.apache.axis2.transport.base,
+                            io.swagger.parser.*; version="${openapitools.swagger.parser.version}",
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/pom.xml]: better to import version range

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+                throw new APISecurityException(APISecurityConstants.API_AUTH_INVALID_CREDENTIALS,
+                        "Invalid JWT token");
+            }
+        }
+
+        // Not found in cache
+        if (!isVerified) {
+            log.debug("Token not found in the cache.");
+            try {
+                payload = new JSONObject(new String(Base64.getUrlDecoder().decode(splitToken[1])));
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java]: posiible array index out of bounds/

Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java]: this split length validation is happening at the OAuthAuthenticator so when we hit this method, it should definitely have an array of length 3

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java (hunk 2 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java]: In the Key manager, we do URL encode. That's why I used URL decode here.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java (hunk 1 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java]: Invalid token length is handled in the OAuthAuthenticator before hitting this code

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java (hunk 3 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java]: Added

File: components/apimgt/org.wso2.carbon.apimgt.gateway/pom.xml (hunk 5 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/pom.xml]: done

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/oauth/OAuthAuthenticator.java [hunk 7]
```diff
@@ -168,7 +179,46 @@ public AuthenticationResponse authenticate(MessageContext synCtx) {
...
+                try {
+                    JSONObject decodedHeader = new JSONObject(new String(Base64.getUrlDecoder()
+                            .decode(apiKey.split("\\.")[0])));
+                    // Check if the decoded header contains type as 'JWT'.
+                    if (APIConstants.JWT.equals(decodedHeader.getString(APIConstants.JwtTokenConstants.TOKEN_TYPE))) {
+                        isJwtToken = true;
+                        if (StringUtils.countMatches(apiKey, APIConstants.DOT) != 2) {
+                            log.debug("Invalid JWT token. The expected token format is <header.payload.signature>");
+                            throw new APISecurityException(APISecurityConstants.API_AUTH_INVALID_CREDENTIALS,
+                                    "Invalid JWT token");
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/oauth/OAuthAuthenticator.java:193]: Cant we state the same debug message here as well?
It provides a hint.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/oauth/OAuthAuthenticator.java [hunk 8]
```diff
@@ -242,6 +292,20 @@ else if(apiVersion == null){
             return new AuthenticationResponse(false, isMandatory, true,
                     APISecurityConstants.API_AUTH_MISSING_CREDENTIALS, "Required OAuth credentials not provided");
         } else {
+            //Start JWT token validation
+            if (isJwtToken) {
+                try {
+                    AuthenticationContext authenticationContext = jwtValidator.authenticate(apiKey, synCtx, openAPI);
+                    APISecurityUtils.setAuthenticationContext(synCtx, authenticationContext, securityContextHeader);
+                    log.debug("User is authorized using JWT token to access the resource.");
```
Comment (isharac) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/oauth/OAuthAuthenticator.java:300]: improve the debug log by appending specific information such as user etc

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+
+    private Cache getCacheFromCacheManager(String cacheName) {
+        return Caching.getCacheManager(
+                APIConstants.API_MANAGER_CACHE_MANAGER).getCache(cacheName);
+    }
+
+    private String getAccessTokenCacheKey(String accessToken, String apiContext, String apiVersion,
+                                          String resourceUri, String httpVerb) {
+        return accessToken + ":" + apiContext + ":" + apiVersion + ":" + resourceUri + ":" + httpVerb;
+    }
```
Comment (shilmyhasan) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java:495]: Constant can be used for ":"

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/oauth/OAuthAuthenticator.java (hunk 7 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/oauth/OAuthAuthenticator.java:193]: Fixed in https://github.com/wso2/carbon-apimgt/pull/9250

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/oauth/OAuthAuthenticator.java (hunk 8 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/oauth/OAuthAuthenticator.java:300]: Fixed in https://github.com/wso2/carbon-apimgt/pull/9250

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java (hunk 9 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java:495]: Fixed in https://github.com/wso2/carbon-apimgt/pull/9250

File: components/apimgt/org.wso2.carbon.apimgt.gateway/pom.xml (hunk 4 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/pom.xml:231]: Added with https://github.com/wso2/carbon-apimgt/pull/8198

**Pull request merged**